Apply vertex attribute fixing to the LOD system for rigged meshes.

fire opened the pull request:
Fixes: https://github.com/godotengine/godot/issues/84479

Here's a sample video with me toggling the mesh lod settings in the background.

[Godot Engine autolod patch rig lod success 02](https://www.youtube.com/watch?v=SjvvWxidm0g)

https://youtu.be/JodufVyd3Ow and https://youtu.be/BMvh1J9ril0

- [ ] Fix crash on same fbx repository.

lyuma commented on scene/resources/3d/importer_mesh.h:
```diff
@@ -82,6 +82,52 @@ class ImporterMesh : public Resource {
 
 	Size2i lightmap_size_hint;
 
+	struct MergedAttribute {
+		Vector3 normal;
+		Vector<int> primary_bone_influence;
+	};
```
This should not have a `Vector<int>` for every vertex. it is inefficient, especially given that there is at most a small constant number of influences per vertex

I think it should be possible with one influence and one weight, maybe two influences at most but I'd like to see a counter-example.
Remember that mesh simplify will try to avoid merging vertices with different values here:

Even if the vertex contains a lot more influences, the one or two most dominant influences should be enough to ensure the vertices are considered different.

fire replied:
The linked issue ‘s glb fails to decimate correctly if we limit to one or two influences.

fire replied:
[test.glb.zip](https://github.com/godotengine/godot/files/14873650/test.glb.zip)

# Limited to 1 influence

https://github.com/godotengine/godot/assets/32321/51234fdb-7bc1-41d5-b4e5-04a47a357e48

# Limited to 8 influences

https://github.com/godotengine/godot/assets/32321/10dc77f3-3d3d-4ed7-8a09-e2cb13bed3b1

fire commented on scene/resources/3d/importer_mesh.cpp:
```diff
@@ -345,35 +347,41 @@ void ImporterMesh::generate_lods(float p_normal_merge_angle, float p_normal_spli
 		const Vector3 *normals_ptr = normals.ptr();
 
-		HashMap<Vector3, LocalVector<Pair<int, int>>> unique_vertices;
+		RBMap<Vector3, LocalVector<Pair<int, int>>> unique_vertices;
```
Restore HashMap?

fire committed:
WIP. Update primary bone influence data structures

Adjust attribute weights and array sizes for LOD generation.

Increase maximum attributes to 32 in rescaleAttributes function
